File: srclib/v3/spss/srclib/putdata.py
# putdata function

import os, re, spss, spssaux, srclib, sys


def putdata(name=None, dir=None, mode=None, dataout=None, impl=None, mult=None):
  msg = None
  if not name:  # name
    msg = "Missing run name."
  else:
    m = re.match(r"^\s*(\w[\w\-]*)\s*$", name)
    if m:
      name = m.group(1)
    else:
      msg = "Run name error."
  if not msg:
    cwd = spssaux.getShow("DIRECTORY")  # spss current working directory
    if not dir:
      path = os.path.join(cwd, name)  # spss current working directory
    else:  # specified directory
      m = re.match(r'^\s*"([~\.\\/:\w\- ]*)"\s*$', dir)   # double quotes
      if m:
        dir = m.group(1).strip()
      else:
        m = re.match(r"^\s*'([~\.\\/:\w\- ]*)'\s*$", dir)   # single quotes
        if m:
          dir = m.group(1).strip()
        else:
          m = re.match(r"^\s*([~\.\\/:\w\-]*)\s*", dir)   # no quotes
          if m:
            dir = m.group(1)
          else:
            msg = "Invalid directory."
      if not msg:
        path = os.path.join(os.path.normpath(dir), name)
  if not msg:  # mode
    if mode:
      m = re.match(r"^\s*(debug|test)\s*$", mode, flags=re.IGNORECASE)
      if m:
        mode = m.group(1).lower()
      else:
        msg = "Mode error."
  if not msg:  # dataout
    if not dataout:
      msg = "Missing dataout."
    else:
      m = re.match(r'^\s*"([~\.\\/:\w\- ]*)"\s*$', dataout)   # double quotes
      if m:
        dataout = m.group(1).strip()
      else:
        m = re.match(r"^\s*'([~\.\\/:\w\- ]*)'\s*$", dataout)   # single quotes
        if m:
          dataout = m.group(1).strip()
        else:
          m = re.match(r"^\s*([~\.\\/:\w\-]*)\s*", dataout)   # no quotes
          if m:
            dataout = m.group(1)
          else:
            msg = "Invalid dataout."
      if not msg:
        if sys.platform.startswith("win"):  # windows
          m = re.match(r"^([a-zA-Z]:|[\\/])", dataout)
        else:  # linux
          m = re.match("^(~|/)", dataout)
        if not m:  # not full path
          dataout = os.path.join(cwd, dataout)  # prefix the current working directory
        dataout = os.path.normpath(dataout)  # normalize the path
  if not msg:  # impl
    if impl:
      m = re.match(r"^\s*(all|\d+)\s*$", impl, flags=re.IGNORECASE)
      if m:
        impl = m.group(1).lower()
      else:
        msg = "Invalid implicate."
  if not msg:  # mult
    if mult:
      m = re.match(r"^\s*(all|\d+)\s*$", mult, flags=re.IGNORECASE)
      if m:
        mult = m.group(1).lower()
      else:
        msg = "Invalid multiple."
  if not msg:
    args = []  # execute putdata
    args.append("/dataout={0}".format(dataout))
    if impl:
      args.append("/impl={0}".format(impl))
    if mult:
      args.append("/mult={0}".format(mult))
    rc = srclib.execute("putdata", path, mode, args)
    if rc != 0:
      msg = "Abnormal termination of putdata"
    else:
      if not os.path.exists(dataout + ".out"):  # write the imputed data
        msg = "Missing " + dataout + ".out file"
      else:
        f = open(dataout + ".out", "r")
        cmd = f.read()
        f.close()
        spss.Submit(cmd)  # execute the command
        if not mode:
          os.remove(dataout + ".imp")
          os.remove(dataout + ".out")
  if msg:
    print msg  # print the error message

File: srclib/v3/R/putdata.R
# putdata function

putdata <- function(name="", dir="", mode="", dataout="", impl="", mult="") {
  msg <<- ""
  path <<- ""
  test <<- ""
  name <<- sub("^\\s+", "", sub("\\s+$", "", name))  # name
  if (identical(name, "")) msg <<- "Missing name."
  else {
    if (!grepl("^[[:alnum:]][[:alnum:]_-]*$", name)) msg <<- "Invalid name."
  }
  if (identical(msg, "")) {  # path
    dir <- sub("^\\s+", "", sub("\\s+$", "", dir))
    if (identical(dir, "")) path <<- name  # no directory specified, use just name
    else {  # specified directory
      if (grepl('^"[^"]*"$', dir)) dir <<- sub('^"\\s*(.*)\\s*"$', "\\1", dir)  # double quotes
      else if (grepl("^'[^']*'$", dir)) dir <<- sub("^'\\s*(.*)\\s*'$", "\\1", dir)  # single quotes
      if (!grepl("^[~\\.\\\\/:[:alnum:]_ -]+$", dir))  msg <<- "Invalid directory."
      else path <<- file.path(dir, name, fsep=.Platform$file.sep)
     }
  }
  if (identical(msg, "")) {  # test
    test <<- sub("^\\s+", "", sub("\\s+$", "", mode))
    if (!identical(test, "")) {
      if (!grepl("^(debug|test)$", test, ignore.case=TRUE)) msg <<- "Invalid test."
    }
  }
  if (identical(msg, "")) {   # dataout
  dataout <<- sub("^\\s+", "", sub("\\s+$", "", dataout))
  if (identical(dataout, "")) msg <<- "Missing dataout."
  else {
    if (grepl('^"[^"]*"$', dataout)) dataout <<- sub('^"\\s*(.*)\\s*"$', "\\1", dataout)  # double quotes
    else if (grepl("^'[^']*'$", dataout)) dataout <<- sub("^'\\s*(.*)\\s*'$", "\\1", dataout)  # single quotes
    if (!grepl("^[~\\.\\\\/:[:alnum:]_ -]+$", dataout))  msg <<- "Invalid dataout."
    }
  }
  if (identical(msg, "")) {   # impl
    impl <<- sub("^\\s+", "", sub("\\s+$", "", impl))
    if (!identical(impl, "")) {
      if (!grepl("^(all|[[:digit:]]+)$", impl, ignore.case=TRUE))  msg <<- "Invalid implicate."
    }
  }
  if (identical(msg, "")) {   # mult
    mult <<- sub("^\\s+", "", sub("\\s+$", "", mult))
    if (!identical(mult, "")) {
      if (!grepl("^(all|[[:digit:]]+)$", mult, ignore.case=TRUE))  msg <<- "Invalid multiple."
    }
  }
  if (identical(msg, "")) {  # execute putdata
    args <- paste("/dataout=", dataout, sep="")
    if (regexpr(" ", args, fixed=TRUE) >= 0) args <- dQuote(args)
    if (!identical(impl, "")) args <- paste(args, " /impl=", impl, sep="")
    if (!identical(mult, "")) args <- paste(args, " /mult=", mult, sep="")
    execute("putdata", args)
    if (identical(msg, "")) {  # write the imputed data
      out <- paste(dataout, "out", sep=".")
      if (file.exists(out)) {
        source(out)
        if (identical(test, "")) {
          file.remove(out)
          file.remove(paste(dataout, "imp", sep="."))
        }
      }
      else  msg <<- "Missing output file."
    }
  }
  if (!identical(msg, "")) {
    cat(msg)  # print the error message
  }
}

